stats/csv/collector.go

```diff
@@ -0,0 +1,246 @@
+	resTags := []string{}
+	ignoredTags := []string{}
+	for tag, flag := range tags {
+		if flag {
+			resTags = append(resTags, tag)
+		} else {
+			ignoredTags = append(ignoredTags, tag)
+		}
+	}
+	sort.Strings(resTags)
+
+	if fname == "" || fname == "-" {
+		logfile := nopCloser{os.Stdout}
+		return &Collector{
+			outfile:     logfile,
+			fname:       "-",
+			resTags:     resTags,
+			ignoredTags: ignoredTags,
+			csvWriter:   csv.NewWriter(logfile),
+		}, nil
+	}
+
+	logfile, err := fs.Create(fname)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Collector{
+		outfile:     logfile,
+		fname:       fname,
+		resTags:     resTags,
+		ignoredTags: ignoredTags,
+		csvWriter:   csv.NewWriter(logfile),
+	}, nil
+}
+
+// Init writes column names to csv file
+func (c *Collector) Init() error {
+	header := MakeHeader(c.resTags)
+	err := c.csvWriter.Write(header)
+	if err != nil {
+		log.WithField("filename", c.fname).Error("CSV: Error writing column names to file")
+	}
+	c.csvWriter.Flush()
+	return nil
+}
+
+// SetRunStatus does nothing
+func (c *Collector) SetRunStatus(status lib.RunStatus) {}
+
+// Run just blocks until the context is done
+func (c *Collector) Run(ctx context.Context) {
+	ticker := time.NewTicker(saveInterval)
+	for {
+		select {
+		case <-ticker.C:
+			c.WriteToFile()
+		case <-ctx.Done():
+			c.WriteToFile()
+			err := c.outfile.Close()
+			if err != nil {
+				log.WithField("filename", c.fname).Error("CSV: Error closing the file")
+			}
+			return
+		}
+	}
+}
+
+// Collect Saves samples to buffer
+func (c *Collector) Collect(scs []stats.SampleContainer) {
+	c.bufferLock.Lock()
+	defer c.bufferLock.Unlock()
+	for _, sc := range scs {
+		c.buffer = append(c.buffer, sc.GetSamples()...)
+	}
+}
+
+// WriteToFile Writes samples to the csv file
+func (c *Collector) WriteToFile() {
+	c.bufferLock.Lock()
+	samples := c.buffer
+	c.buffer = nil
+	c.bufferLock.Unlock()
+
+	if len(samples) > 0 {
+		c.csvLock.Lock()
+		defer c.csvLock.Unlock()
+		for _, sc := range samples {
+			for _, sample := range sc.GetSamples() {
+				func(sample stats.Sample) {
+					row := SampleToRow(&sample, c.resTags, c.ignoredTags)
+					err := c.csvWriter.Write(row)
+					if err != nil {
+						log.WithField("filename", c.fname).Error("CSV: Error writing to file")
+					}
+				}(sample)
+			}
+		}
+		c.csvWriter.Flush()
+	}
+}
+
+// Link returns a dummy string, it's only included to satisfy the lib.Collector interface
+func (c *Collector) Link() string {
+	return ""
+}
+
+// MakeHeader creates list of column names for csv file
+func MakeHeader(tags []string) []string {
+	tags = append(tags, "extra_tags")
+	return append([]string{"metric_name", "timestamp", "metric_value"}, tags...)
+}
+
+// SampleToRow converts sample into array of strings
+func SampleToRow(sample *stats.Sample, resTags []string, ignoredTags []string) []string {
+	if sample == nil {
+		return nil
+	}
+
+	row := []string{}
```

Comment:
You do know how many columns you will have (3 + len(restags) + 1 for the the extra tags. You can do `row := make([]string, 0, 3 + len(resTags) + 1)` and not change any of the other code.

Reply:
On similar note you can probably reuse the same slice over and over again as you are populating it and than writing it. This will have a lot of performance gain IMHO as it will practically remove a big chunk of allocation